keps/83-workload-preemption/README.md

```diff
@@ -279,19 +261,18 @@ The algorithm is like follows:
     2. [New step] if no flavor was found, use the first flavor in the list that
        has more min quota than the Workload request.
 
-We will call the flavor assignments type 0, 1 or 2, depending on which step they
-were found.
-    
-A flavor assignment of type 1 means that we need to preempt or wait for other
-workloads in the cohort to finish to accommodate this workload, because the
-ClusterQueue is lending its resources.
-
-A flavor assigment of type 2 means that we need to preempt or wait for other
-workloads in the ClusterQueue to finish to accomodate this workload.
+Some highlights:
+- A Workload could get flavor assignments at different steps for different
+  resources.
+- Assigments that require preemption implicitly do not borrow quota.
```

Comment:
I think it is reasonable to support the case where a higher priority workload can preempt lower priority ones that are coming from any CQ with usage above its min. We can leave that as a followup, but I think we should make sure that the initial implementation doesn't block this case.

Reply:
> I think it is reasonable to support the case where a higher priority workload can preempt lower priority ones that are coming from any CQ with usage above its min. 

Isn't that what we are already proposing? Or do you mean that the CQ of the pending workload would go above its min by preempting other workloads on CQs that aren't above their min?

Reply:
The statement "Assigments that require preemption implicitly do not borrow quota." implies that preemption is allowed only if the pending workload can fit under its CQ quota after preemption.

Consider the case where all admitted workloads in CQ1 are of equal or higher priority than the pending workload, and there are lower priority workloads from CQ2 that are admitted via borrowed quota from CQ3, then I am suggesting that the pending workload from CQ1 should preempt the lower priority ones from CQ2 to release the unused resources of CQ3.

Reply:
Gotcha....

I think the implementation in this proposal can still be extended to support that:
we query what is the minimum priority among the admitted ClusterQueues in the flavor and see if it's higher than or equal to the pending Workload, then choose any flavor, but mark it as borrowing. Then run the preemption algorithm considering all workloads in the cohort.

Reply:
sounds good, lets document this case then

Reply:
Actually, to accommodate for that possible expansion, I renamed the enums a little, to include the fact that we are limiting preemption to situations of "reclaiming quota". Please see update.